activate

```diff
@@ -1,95 +1,218 @@
-if [ -e "${ACTIVATEDFILE}" ]; then
-    echo "${ACTIVATEDFILE} exists, doing nothing"
-    exit 0
-fi
-
-SCRIPT="${EOS_PHONE_HOME_ACTIVATE_URL:-https://home.endlessm.com/v1/activate}"
-
-image=$(attr -q -g eos-image-version /sysroot 2>/dev/null ) || true
-if [ "${image}" != "" ]; then
-    image_root=$(attr -q -g eos-image-version / 2>/dev/null ) || true
-    if [ "${image_root}" != "" ]; then
-        image="${image_root}"
-    fi
-fi
-
-release=$(cat /etc/os-release | grep ^VERSION= | head -n1 | awk -F= '{ print $2 }') || true
-
-vendor=""
-product=""
-serial=""
-if [ -d "/sys/class/dmi/id" ]; then
-    vendor=$(< /sys/class/dmi/id/sys_vendor) || true
-    product=$(< /sys/class/dmi/id/product_name ) || true
-    serial=$(< /sys/class/dmi/id/product_serial) || true
-else
-    compat=$(cut -d '' -f1 < /proc/device-tree/compatible) || true
-    vendor=$(echo ${compat} | cut -d, -f1)
-    product=$(echo ${compat} | cut -d, -f2)
-    serial=$(< /sys/class/endless_mfgdata/entries/SSN) || serial=''
-fi
-
-# trim whitespace from strings
-vendor="${vendor%"${vendor##*[![:space:]]}"}"
-product="${product%"${product##*[![:space:]]}"}"
-serial="${serial%"${serial##*[![:space:]]}"}"
-
-# detect live system
-live="false"
-if grep -qw "endless.live" /proc/cmdline; then
-    image="true"
-fi
-
-# format as JSON
-vars="image release vendor product serial live"
-export ${vars}
-json="$(python -c 'import json, os, sys; print json.dumps(dict((v,os.environ[v]) for v in sys.argv[1:]))' ${vars})"
-
-# report in
-if ! out=$(wget -O - -q --method=PUT --body-data="${json}" "${SCRIPT}") ; then
-    echo "failed PUT to $SCRIPT with data: ${json}"
-    exit 1
-fi
-if ! ( echo "${out}" | python -c "import json, sys; sys.exit(0 if json.load(sys.stdin)['success'] == 'true' else 1)" 2>/dev/null ); then
-    echo "didn't get success from PUT to ${SCRIPT} with data: ${json}"
-    echo "got response: ${out}"
-    exit 1
-fi
-
-# touch file to record activation
-echo > "${ACTIVATEDFILE}"
+#!/usr/bin/env python3
 
+import json
+import os
+import re
+import subprocess
+import sys
+import urllib.request
+
+class Activation(object):
+    ACTIVATION_ENDPOINT = os.getenv('EOS_PHONE_HOME_ACTIVATE_URL', "https://home.endlessm.com/v1/activate")
+    STATE_DIRECTORY = "/var/lib/eos-phone-home"
+
+    ACTIVATION_VARIABLES = { 'live': 'is_live',
+                             'image': 'image',
+                             'release': 'release',
+                             'vendor': 'vendor',
+                             'product': 'product',
+                             'serial': 'serial' }
+
+    def _get_vendor(self):
+        vendor = None
+
+        try:
+            if os.path.isdir('/sys/class/dmi/id'):
+                with open('/sys/class/dmi/id/sys_vendor') as vendor_file:
+                    vendor = vendor_file.read().strip()
+            else:
+                dt_data = None
+                with open('/proc/device-tree/compatible') as dt_file:
```

Comment:
It's easy to miss in the shell version (!) but /proc/device-tree/compatible is a list of NUL-separated vendor,product strings. Thanks to the unique way that ARM Linux works, we don't know which of the possible strings are "us" because the bootloader just picks the DT file based on boot-time knowledge which isn't passed on to us, but the first string in the file is the most common/preferred device. So we just make do with that.